pkg/git/git.go

```diff
@@ -0,0 +1,172 @@
+	}
+	if !ok {
+		// nothing to do.
+		return nil
+	}
+	args := []string{
+		"--git-dir", filepath.Join(g.Location, ".git"),
+		"--work-tree", g.Location,
+		"commit",
+		"-m",
+		"Push changes to remote",
+		g.Filename,
+	}
+	if err := g.runCmd(gitCmd, args...); err != nil {
+		return fmt.Errorf("failed to run commit: %w", err)
+	}
+	return nil
+}
+
+// CreateBranch creates a branch if it differs from the base.
+func (g *CLIGit) CreateBranch() error {
+	if g.Base == g.Branch {
+		return nil
+	}
+	fmt.Println("creating new branch")
+	args := []string{
+		"--git-dir", filepath.Join(g.Location, ".git"),
+		"--work-tree", g.Location,
+		"checkout",
+		"-b",
+		g.Branch,
+	}
+	if err := g.runCmd(gitCmd, args...); err != nil {
+		return fmt.Errorf("failed to create new branch %s: %w", g.Branch, err)
+	}
+	return nil
+}
+
+// CreateRepository bootstraps a plain repository at a given location.
+func (g *CLIGit) CreateRepository() error {
+	return errors.New("implement me")
+}
+
+// IsRepository returns whether a location is a git repository or not.
+func (g *CLIGit) IsRepository() error {
+	// Note that this is redundant in case of CLI git, because the git command line utility
+	// already checks if the given location is a repository or not. Never the less we do this
+	// for posterity.
+	if _, err := os.Stat(filepath.Join(g.Location, ".git")); os.IsNotExist(err) {
+		return err
+	}
+	return nil
+}
+
+// HasChanges returns whether a location has uncommitted changes or not.
+func (g *CLIGit) HasChanges() (bool, error) {
+	args := []string{
+		"--git-dir", filepath.Join(g.Location, ".git"),
+		"--work-tree", g.Location,
+		"status",
+		"-s",
+	}
+	out, err := g.Runner.Run(gitCmd, args...)
+	if err != nil {
+		return false, fmt.Errorf("failed to check if there are changes: %w", err)
+	}
+	return string(out) != "", nil
+}
+
+// Push will push to a remote.
+func (g *CLIGit) Push() error {
+	fmt.Println("pushing to remote")
+	args := []string{
+		"--git-dir", filepath.Join(g.Location, ".git"),
+		"--work-tree", g.Location,
+		"push",
+		g.Remote,
+		g.Branch,
+	}
+	if err := g.runCmd(gitCmd, args...); err != nil {
+		return fmt.Errorf("failed to push changes to remote %s with branch %s: %w", g.Remote, g.Branch, err)
+	}
+	return nil
+}
+
+// Add will add all unstaged changes.
+func (g *CLIGit) Add() error {
+	fmt.Println("adding unstaged changes")
+	args := []string{
+		"--git-dir", filepath.Join(g.Location, ".git"),
+		"--work-tree", g.Location,
+		"add",
+		g.Filename,
+	}
+	if err := g.runCmd(gitCmd, args...); err != nil {
+		return fmt.Errorf("failed to run add: %w", err)
+	}
+	return nil
+}
+
+// runCmd is a convenient wrapper around running commands with error output when the output is not needed but needs to
+// be logged.
+func (g *CLIGit) runCmd(c string, args ...string) error {
+	out, err := g.Runner.Run(c, args...)
+	if err != nil {
+		fmt.Printf("failed to run cmd %s with output: %s\n", c, string(out))
+	}
+	return err
+}
```

Comment:
is the first arg ever not `gitCmd`? If so we could remove the arg and rename `runGitCmd`

Reply:
Hmmm, good point, good point. I guess it can be, yes.